File: scripts/slices.lua
-- Copyright (C) 2018  David Capello
--
-- This file is released under the terms of the MIT license.
-- Read LICENSE.txt for more information.

do
  local s = Sprite(32, 32)

  local a = s:newSlice()
  local b = s:newSlice(0, 2, 8, 10)
  local c = s:newSlice{ x=0, y=0, width=32, height=32 }
  assert(a.bounds == nil)
  assert(b.bounds == Rectangle(0, 2, 8, 10))
  assert(c.bounds == Rectangle(0, 0, 32, 32))

  s:deleteSlice(b)
  assert(a == s.slices[1])
  assert(c == s.slices[2])
end

File: src/doc/slices.cpp
// Aseprite Document Library
// Copyright (c) 2017 David Capello
//
// This file is released under the terms of the MIT license.
// Read LICENSE.txt for more information.

#ifdef HAVE_CONFIG_H
  #include "config.h"
#endif

#include "doc/slices.h"

#include "base/debug.h"
#include "doc/slice.h"

#include <algorithm>

namespace doc {

Slices::Slices(Sprite* sprite) : m_sprite(sprite)
{
}

Slices::~Slices()
{
  for (Slice* slice : m_slices) {
    slice->setOwner(nullptr);
    delete slice;
  }
}

void Slices::add(Slice* slice)
{
  m_slices.push_back(slice);
  slice->setOwner(this);
}

void Slices::remove(Slice* slice)
{
  auto it = std::find(m_slices.begin(), m_slices.end(), slice);
  ASSERT(it != m_slices.end());
  if (it != m_slices.end())
    m_slices.erase(it);

  slice->setOwner(nullptr);
}

Slice* Slices::getByName(const std::string& name) const
{
  for (Slice* slice : *this) {
    if (slice->name() == name)
      return slice;
  }
  return nullptr;
}

Slice* Slices::getById(ObjectId id) const
{
  for (Slice* slice : *this) {
    if (slice->id() == id)
      return slice;
  }
  return nullptr;
}

} // namespace doc
